src/mlpack/base.hpp

```diff
@@ -68,29 +70,6 @@
   #define mlpack_force_inline __forceinline
 #endif
 
-// detect C++17 mode
-#if (__cplusplus >= 201703L)
-  #undef  MLPACK_HAVE_CXX17
-  #define MLPACK_HAVE_CXX17
-#endif
-
-#if defined(_MSVC_LANG)
-  #if (_MSVC_LANG >= 201703L)
-    #undef  MLPACK_HAVE_CXX17
-    #define MLPACK_HAVE_CXX17
-  #endif
-#endif
-
-#if defined(MLPACK_HAVE_CXX17)
-  #include <any>
-  #include <string_view>
-  #define MLPACK_ANY std::any
-  #define MLPACK_ANY_CAST std::any_cast
-  #define MLPACK_STRING_VIEW std::string_view
-#elif defined(_MSC_VER)
-  #error "When using Visual Studio, mlpack should be compiled with /std:c++17 or newer."
-#endif
-
```

Comment:
Suggest to keep the C++17 detection parts (directly related to `__cplusplus >= 201703L` and `_MSVC_LANG >= 201703L), so that we end up with `MLPACK_HAVE_CXX17` defined.  This is because MSVC likes to be "special".

Suggest to change the `#if defined(MLPACK_HAVE_CXX17)` block to:
```
#if !defined(MLPACK_HAVE_CXX17)
  #error "Need to enable C++17 mode in your compiler"
#endif
```